### PR TITLE
feat(open): modify `devServer.options.open` to take an object

### PR DESCRIPTION
**Pull request by dangreenisrael**
This PR updates `devServer.options` to take an array, making it consistent with the underlying library, open

https://www.npmjs.com/package/open

- [ ] This is a **bugfix**
- [x] This is a **feature**
- [ ] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

I didn't see any tests for the unmodified code, I did however update the `options.json` schema

### Motivation / Use-Case

This updates `devServer.options` to match the underlying library that is used in the implementation.

https://github.com/webpack/webpack-dev-server/issues/1769

### Breaking Changes

N/A
### Additional Info
N/A

### CODE REVIEW

**Comment by dangreenisrael** [comment]
Hey @hiroppy. I added some tests for the config, but I didn't see anything in there testing the `status` file in general. Did you want me to add a suite for it, or did I miss something?

Thanks for helping maintain such an awesome project!!!

**Comment by dangreenisrael** [comment]
Hey @evilebottnawi. We wound up not auto-open in the end so I'm not able to do it at work, and I don't have much extra time right now (I've got a 14 month old). That said, my plan is to try to get the tests written soon so this can land.

Sorry for the delays

File: test/Status.test.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+const catchMock = jest.fn(() => 'foo');
+open.mockReturnValue({
+  catch: catchMock,
+});
+
+const uri = 'http://127.0.0.1';
+const log = {
+  info: jest.fn(),
+  warn: jest.fn(),
+};
```
Comment (evilebottnawi) [inline test/Status.test.js]: Can we use `just.spyOn` to better reading tests, thanks!

Comment (dangreenisrael) [reply test/Status.test.js]: We need to create `log` anyway.  Would you prefer:

```javascript
const log = {
  info: () => {}
  warn: () => {}
}

const infoSpy = jest.spyOn(log, 'info')
const warnSpy = jest.spyOn(log, 'warn')
```

File: lib/utils/status.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ function status(uri, options, log, useColor) {
     let openOptions = {};
     let openMessage = 'Unable to open browser';
 
-    if (typeof options.open === 'string') {
+    if (typeof options.open === 'string' || Array.isArray(options.open)) {
```
Comment (evilebottnawi) [inline lib/utils/status.js]: hm, why is array, based on docs https://github.com/sindresorhus/open it should be object

File: lib/options.json [hunk 3]
```diff
@@ -241,6 +241,12 @@
         },
         {
           "type": "boolean"
+        },
+        {
+          "type": "array",
+          "items": {
+            "type": "string"
```
Comment (evilebottnawi) [inline lib/options.json]: items, can be not only strings

File: lib/utils/status.js (hunk 2 above)
Comment (dangreenisrael) [reply lib/utils/status.js]: Options is an object, which takes `app`. App takes either a string or an array. https://www.npmjs.com/package/open#usage . This keeps with the existing API of this project, passing only the `app` value for `open` in options.

i.e:
```
{
  open: "google chrome"
}
```

and

```
{
  open: ['google chrome', '--incognito']
}
```

File: lib/options.json (hunk 3 above)
Comment (dangreenisrael) [reply lib/options.json]: According to this, I believe they are only string https://www.npmjs.com/package/open#usage

**Comment by dangreenisrael** [comment]
> I think we should do this in other way.
> 
> * Let's make `open` will be `Boolean`, `String` or `Object`
> 
> ```
> if (Object.isObject(this.open)) {
>   openOptions = Object.assign({}, options.open);
> }
> ```

I'm pretty sure the value of  open's `options.app` which is what we are currently accepting in our `options.open`can be either `Boolean`, `String`, or `Array<String>`

File: lib/options.json (hunk 3 above)
Comment (evilebottnawi) [reply lib/options.json]: `{wait: true}`

File: lib/utils/status.js (hunk 2 above)
Comment (evilebottnawi) [reply lib/utils/status.js]: Developers can want to use other options example: `{wait: true}`, also in future package can accept more arguments

Comment (dangreenisrael) [reply lib/utils/status.js]: Good call @evilebottnawi

Comment (dangreenisrael) [reply lib/utils/status.js]: I should have some time tomorrow night to work on this again

File: test/Status.test.js (hunk 1 above)
Comment (dangreenisrael) [reply test/Status.test.js]: ???

**Comment by dangreenisrael** [comment]
@evilebottnawi Based on what you requested (or my understanding of it), I made the following update:


Allow for 3 options to pass as options.open:

- Boolean: runs without any options and thus uses the default browser
- String: opens with the browser of your choice
- Object: takes a settings object as defined by https://github.com/sindresorhus/open

As we are now doing now allowing for more complex (and possibly confusing) options to be passed in I have added logging aimed at making this more discoverable to a user.

If they pass in an object with neither `app` nor `wait` OR something that is not ( boolean | string | object ) they will get a warning telling them that they passed in an invalid config and a link to the `open` docs.

With that, I have also hard coded the open failure message, as config related info should now be caught earlier

**Comment by dangreenisrael** [comment]
If you guys are +1 to this I’ll update the tests and relent docs

**Comment by dangreenisrael** [comment]
Bump

**Comment by hiroppy** [comment]
CI fails

```
Error: expect(jest.fn()).toBeCalledWith(expected)

Expected mock function to have been called with:
  {"app": ["Google Chrome", "incognito"]}
as argument 2, but it was called with
  ["Google Chrome", "incognito"].

Difference:

  Comparing two different types of values. Expected object but received array.
    at Object.toBeCalledWith (/Users/vsts/agent/2.150.3/work/1/s/test/Status.test.js:58:20)
    at Object.asyncJestTest (/Users/vsts/agent/2.150.3/work/1/s/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at e (/Users/vsts/agent/2.150.3/work/1/s/node_modules/jest-jasmine2/build/queueRunner.js:43:12)
    at mapper (/Users/vsts/agent/2.150.3/work/1/s/node_modules/jest-jasmine2/build/queueRunner.js:26:19)
    at promise.then (/Users/vsts/agent/2.150.3/work/1/s/node_modules/jest-jasmine2/build/queueRunner.js:73:41)
```

**Comment by dangreenisrael** [comment]
I just wanted a +1 to the approach before I took the time to update the tests. I should have some time to get it done this weekend

**Comment by evilebottnawi** [comment]
@dangreenisrael +1

**Comment by hiroppy** [comment]
@dangreenisrael friendly ping

**Comment by dangreenisrael** [comment]
I haven't forgotten about you guys. Just super busy. 

Thanks for the friendly ping tho

**Comment by hiroppy** [comment]
@evilebottnawi I updated this pr.

I'll do below after merging.

- modify `test/status`
- open has many keys
  - https://github.com/sindresorhus/open#options
  - we consider accepting an object or array

**Comment by hiroppy** [comment]
need to update documentation

File: lib/utils/runOpen.js [hunk 4]
```diff
@@ -2,18 +2,29 @@
...
+}
+
 function runOpen(uri, options, log) {
   let openOptions = {};
-  let openMessage = 'Unable to open browser';
 
   if (typeof options.open === 'string') {
     openOptions = { app: options.open };
-    openMessage += `: ${options.open}`;
+  } else if (typeof options.open === 'object') {
```
[the remark this answers is not in the document]
Comment (dangreenisrael) [reply lib/utils/runOpen.js]: https://github.com/webpack/webpack-dev-server/pull/1770#pullrequestreview-229502805

**Comment by dangreenisrael** [comment]
Is this still something people are still interested in. If so, can we agree on an interface in principle before any more code changes?

File: test/Status.test.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    });
+    it('should call open with app:["app_name", "option"] is open is an array ', () => {
+      const options = {
+        open: { app: ['Google Chrome', 'incognito'] },
+      };
+      status(uri, options, log, true);
+      expect(open).toBeCalledWith(uri, { app: ['Google Chrome', 'incognito'] });
+    });
+  });
+});
```
Comment (evilebottnawi) [inline test/Status.test.js]: Please move this tests to https://github.com/webpack/webpack-dev-server/blob/master/test/server/open-option.test.js

File: lib/utils/runOpen.js [hunk 6]
```diff
@@ -2,18 +2,29 @@
...
+    openOptions = options.open;
+    if (!options.app && !options.wait) {
+      handleInvalidOptionsConfig(log, options.open);
+    }
```
Comment (evilebottnawi) [inline lib/utils/runOpen.js]: Why we need this? open package can add more option in future

**Comment by dangreenisrael** [comment]
Wow, totally forgot about this. I can try to look into in the next few weeks. If someone else wants to take it over I promise I won't be offended

**Comment by stefan-winkler-diconium** [comment]
@dangreenisrael friendly ping - It would be super-useful to finally have proper "open" support for webpack. Do you think you can wrap this up soon?

**Comment by evilebottnawi** [comment]
@stefan-winkler-diconium you can send another PR and fix problems described here

**Comment by dangreenisrael** [comment]
I've got a broken thumb, so I won't be doing any non-work required coding for at least a month. Sorry

**Pull request closed without merging**